sdk/core/src/main/java/com/klaviyo/coresdk/networking/KlaviyoProperties.kt

```diff
@@ -0,0 +1,112 @@
+package com.klaviyo.coresdk.networking
+
+import com.klaviyo.coresdk.utils.KlaviyoPreferenceUtils
+import java.io.Serializable
+
+/**
+ * Abstract class that wraps around a map to control access to its contents.
+ * Provides helper functions to add properties to the map while controlling the keys available for entry
+ */
+abstract class KlaviyoProperties {
+    internal val propertyMap: MutableMap<String, Serializable> = mutableMapOf()
+
+    /**
+     * Adds a new key/value pair to the map.
+     * [KlaviyoPropertyKeys] adds some control to what keys our property maps recognise
+     */
+    fun addProperty(propertyKey: KlaviyoPropertyKeys, value: String) {
+        propertyMap[propertyKey.name] = value
+    }
+
+    /**
+     * Adds a custom property to the map.
+     * Custom properties can define any key name that isn't already reserved by Klaviyo
+     */
+    abstract fun addCustomProperty(key: String, value: Serializable)
+
+    /**
+     * Fetches and returns our property map after converting it into a immutable map
+     */
+    internal fun toMap(): Map<Any, Any> {
+        return propertyMap.toMap()
+    }
+}
+
+/**
+ * Controls the data that can be input into a map of customer properties recognised by Klaviyo
+ */
+class KlaviyoCustomerProperties: KlaviyoProperties() {
+
+    init {
+        setAnonymousId()
+        setDefaultEmail()
```

Comment:
you have to have an email?

Reply:
Ya, actually an idea I stole from the iOS SDK. Its cause of the automatic token generation, it could happen before the user assigns any email address, so the token will upload to Klaviyo as part of an anonymous profile. Putting this call here will see if we have an email address cached already from a previous request and add it to the request so that it associates to a real profile and reduces the creation of anonymous profiles

Reply:
As an alternative to this, I can later add the manual token generation path, which will put the time the token is generated and sent off to Klaviyo more in the user's hands with a single API call